Components: Add onFocusOutside replacement to Popover onClickOutside

aduth opened the pull request:
Fixes #14754, fixes #14849.

This pull request seeks to refactor popovers to avoid relying on "click outside" behaviors in popovers, deprecating the `Popover` component's `onClickOutside` prop in favor of a substitute `onFocusOutside`. This accounts for more reasons for focus transitions not previously reflected in `onClickOutside`, such as that described by the bug of #14754.

**Status:** While functionally complete, I'm considering this to be blocked by an apparent incompatibility with a tangentially-related bug described at #14849 . Notably, when toggling an option in the More Options menu, the menu will no longer become closed again from clicks elsewhere in the page.

**Testing instructions:**

Repeat Steps to Reproduce from #14754 , verifying that the menu is closed as expected.

Verify there are no changes in behavior from other Popover usage, particularly specialized cases such as:

- Clicking a DropdownMenu toggle button should toggle itself closed when currently opened
- Clicking a result in the link suggestions autocomplete does not dismiss the URL popover

aduth committed:
Components: Add onFocusOutside alternative to Popover onClickOutside

aduth committed:
Components: Refactor Dropdown to use onFocusOutside

aduth committed:
Format Library: Refactor inline link to use onFocusOutside

gziolo commented:
> Clicking a result in the link suggestions autocomplete does not dismiss the URL popover

There seems to be a different behavior when using mouse and keyboard. Well, there might be something wrong with the URL popover and autocomplete:
- when I use an arrow key, select one of the results and press `space` it doesn't get selected
- when I use an arrow key, select one of the results and press `enter` it gets selected and focus goes back to RichText field

gziolo commented on packages/components/src/popover/detect-outside.js:
```diff
@@ -1,24 +1,21 @@
- * External dependencies
+ * WordPress dependencies
  */
-import clickOutside from 'react-click-outside';
```
Unrelated question to this PR. Can we also refactor `Modal` component to stop using `react-click-outside`? :)

aduth replied:
> Unrelated question to this PR. Can we also refactor `Modal` component to stop using `react-click-outside`? :)

I would like to, yes.

( https://github.com/WordPress/gutenberg/pull/6261#discussion_r191758392 😞 )

gziolo replied:
There is still hope then :)

gziolo replied:
https://unpkg.com/react-click-outside@3.0.1/index.js

Well, it looks like `react-click-outside` is a simplified version of our own HOC :)

gziolo replied:
I opened a follow-up to address it – #16878.

gziolo commented on packages/components/src/popover/index.js:
```diff
@@ -94,6 +96,47 @@ class Popover extends Component {
+	 *
+	 * @param {FocusEvent} event Focus event from onFocusOutside.
+	 */
+	onFocusOutside( event ) {
```
How can I test this shim? By updating one of the places with usage?

aduth replied:
You could locally revert one or both of 55ca76a5427876ab91e6966272e973fe295e0275 or 3e8ffc4673db9f9527173d03b5018b53d7c78a9d to restore those components' use of `onClickOutside`.

gziolo reviewed (commented):
I will finish this review tomorrow. It's looking great so far. I left some questions.

aduth commented:
This one was a pain to rebase, in good part because of #15053 (cc @youknowriad). 

Worth noting that this will resolve some lingering errors after #11360 logged frequently in StrictMode concerning `findDOMNode` from `react-click-outside` (related: https://github.com/kentor/react-click-outside/issues/45). Also related to prior conversation https://github.com/WordPress/gutenberg/pull/14851#discussion_r273979181.

youknowriad commented on packages/components/src/popover/detect-outside.js:
```diff
@@ -1,24 +1,21 @@
 }
 
-export default clickOutside( PopoverDetectOutside );
+export default withFocusOutside( PopoverDetectOutside );
```
This could be a nice hook const `useOnFocusOutside( ref, handler )`

noisysocks reviewed (approved):
I rebased this PR to fix some conflicts that existed in the `CHANGELOG.md` files. You may want to double check that the changes to `CHANGELOG.md` make sense.

I followed the testing steps in the description and everything seems okay. I also played with some `Popover` and `DropdownMenu` components and couldn't see anything off. I was not able to follow the original steps in the linked issue as that issue was fixed via a different approach.

Would love to get this in as it fixes a bug that's blocking #16582.

aduth commented:
Thanks @noisysocks for the assist here. Looking again over the code and with some final testing, I think this is in good shape for merge.

However, I might like to solicit some accessibility feedback here. The changes seem like a reasonable improvement to avoid assumptions of pointer devices, but I'm curious if there are edge cases I've overlooked. Since the associated issue #14754 has been addressed in a separate pull request, I think it should be safe to spare some additional time before we commit to modifying the API.

aduth commented:
Thanks @noisysocks for the assist here. Looking again over the code and with some final testing, I think this is in good shape for merge. 👍

aduth commented:
Hmm, the end-to-end test failures may be legitimate.

aduth commented:
The end-to-end tests seem to be getting stuck when trying to switch from "Code" editor back to "Visual" in the [container blocks test cases](https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/specs/plugins/container-blocks.test.js). I can't see any reason why it would be such a specific issue here.

My hunches are:

- Something specific with the [plugin under test](https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/plugins/inner-blocks-templates/index.js) and its template locking which is causing trouble?
- The implementation of [`withFocusOutside` using timeouts](https://github.com/WordPress/gutenberg/blob/d45c8ea9f3adafba76676a29f07f28dfd0f2d444/packages/components/src/higher-order/with-focus-outside/index.js#L81-L85) could be flawed and allow for race conditions

Separately, I stumbled upon this code, which seems like something we might want to update:

https://github.com/WordPress/gutenberg/blob/d45c8ea9f3adafba76676a29f07f28dfd0f2d444/packages/block-library/src/image/edit.js#L208

noisysocks commented:
I figured out how to recreate the bug with some simple steps, but I'm no closer to figuring out what's causing it:

1. Create a new post
2. Open the _More tools & options_ menu
3. Click any menu item, e.g. enable _Spotlight mode_
4. Click outside of the menu to dismiss it. The menu does not close

noisysocks commented:
Figured it out! It's because clicking on the menu item causes a focus loss. This was reported in https://github.com/WordPress/gutenberg/issues/14849. I'll include a fix here.

noisysocks committed:
MenuItem: Always use an IconButton component so as to avoid focus loss.

Switching between `Button` and `IconButton` causes React to remount the
`MenuItem` component. This causes a focus loss as well as E2E failures.

There's no need to use a `Button` for `MenuItems` that are unselected,
since we can simply pass `icon={ undefined }` to the `IconButton`.

noisysocks commented:
I pushed up 738d987 which fixes the E2E test failures and fixes #14849. 

Would you mind giving this a quick test, @gziolo?

---

**Testing instructions:**

Verify that #14849 is fixed:

1. Navigate to Posts > Add New
2. Click or otherwise activate the "More Options" menu in the top-right
3. Press <kbd>Spacebar</kbd> (assumes first menu item focused)
4. Note focus loss

Verify there are no changes in behavior from other Popover usage, particularly specialized cases such as:

- Clicking a DropdownMenu toggle button should toggle itself closed when currently opened
- Clicking a result in the link suggestions autocomplete does not dismiss the URL popover

gziolo commented on packages/components/src/menu-item/index.js:
```diff
@@ -47,32 +46,26 @@ export function MenuItem( {
-		tagName,
-		{
+	return (
+		<IconButton
```
It's a bit stretched use of IconButton in the case when there is no icon 😅 

However, I can't think of another simple way of handling it where this doesn't trigger re-render of the button. I noticed that this behavior is broken only because we swap `IconButton` with `Button` for no reason when only the icon gets removed/added.

noisysocks replied:
Yep, exactly! Switching between the two components causes React to remount the component and lose focus. Looking at `IconButton`, it seemed to me that `icon={}` is an optional prop so I felt comfortable doing it this way.

youknowriad replied:
One thing we keep discussing here and there is why not just add an `icon` prop to the `Button` block and deprecate the `IconButton` (or make it just a shortcut)

gziolo replied:
I'm sure @drw158 @cburton4 have some good ideas on how to approach it :)

I'm fine with anything which makes it less confusing. I guess having more flexible `Button` and `IconButton` as a backward-compatible alias makes a lot of sense.

drw158 replied:
We have an issue open to discuss combining IconButton and Button in #16541 

We are leaning towards keeping them separate, but I'll look at this closer to understand the disadvantage.

drw158 replied:
I'm assuming from the code that this is an issue because Button does not have an icon option. Switching between icon and non-icon buttons for Menu Items is causing technical problems.

Could this be solved by adding an icon option to the Button component, but still keep IconButton a separate component?

youknowriad replied:
To be fair, we already do the opposite here. Not pass the "icon" prop to the IconButton (so it behaves like a Button).  If it doesn't make sense to merge the two components into one, I feel we shouldn't do anything specific about it.

gziolo commented:
@afercia, would you mind running some sanity check for the `Manage All Reusable Blocks` link in the menu. I have issues confirming that it gets announced as link as expected. I outputted its HTML content and it is still an anchor with role `menuitem` as expected but I can't confirm it with VoiceOVer.

Aside: yes, I still keep it in mind that we should have no links in the dropdown menu :) I need to find a related issue where @mapk proposes we move it to the Manage Blocks modal and tackle it at last.

gziolo commented:
This PR will also partially resolve https://github.com/WordPress/gutenberg/issues/15501. There is a remaining issue with focus loss when switching between Code and Visual editor when the post is empty.

afercia commented:
Thanks for working on this! @gziolo I tested a bit and seems to me everything works as expected. Just one small thing: should the small horizontal "jump" of menuitems with the check icon be fixed in this PR? See animated GIF:

![jump](https://user-images.githubusercontent.com/1682452/62293383-c8cd3180-b468-11e9-877a-53f095d86fea.gif)

Regarding the way screen readers announce `Manage All Reusable Blocks`: it has a `menuitem` role so it's announced as a "menu button" (or similar, depending on the browser / screen reader). ARIA roles override the native role.

However, it still creates some issues. For example: in VoiceOver, this menuitem is not listed in the "Form Controls" list while the other ones are:

<img width="1141" alt="Screenshot 2019-08-01 at 13 40 40" src="https://user-images.githubusercontent.com/1682452/62293592-3ed19880-b469-11e9-82cb-a17bab4ceefd.png">

- open the More menu
- press Command Option U to open the VoiceOver "rotor"
- right / left arrows to go to the "Form Controls" panel
- down / up arrows to browse the list of form controls
- see that `Manage All Reusable Blocks` is missing

For clarity, menuitems can be links. However, an ARIA menu is supposed to be _either_ a menu of actions _or_ a navigation menu. When it's a navigation menu, it should be wrapped in a `<nav>` element, see the W3C example:

Menu or Menu bar design pattern:
https://www.w3.org/TR/wai-aria-practices-1.1/#menu

Navigation Menubar Example: 
https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html

> Since the menubar presents a site navigation system, it is wrapped in a navigation region implemented with a `nav` element that has an `aria-label` that matches the label on the menubar.

The point is that an ARIA menu _can't be_ a mix of the two things. In this case, all the menuitems perform an action in the current page except `Manage All Reusable Blocks`, which triggers navigation to another page. This is unexpected for users and there's no way to communicate it properly. See #13390

gziolo commented:
> Thanks for working on this! @gziolo I tested a bit and seems to me everything works as expected. Just one small thing: should the small horizontal "jump" of menuitems with the check icon be fixed in this PR?

Good catch, it needs to be fixed, it would be a regression.

noisysocks commented:
Thanks for testing all!

> The point is that an ARIA menu can't be a mix of the two things. In this case, all the menuitems perform an action in the current page except Manage All Reusable Blocks, which triggers navigation to another page. This is unexpected for users and there's no way to communicate it properly. See #13390

Thanks for going into this. Let's address the issues with _Manage All Reusable Blocks_ as part of #13390. I agree we should remove the menu item altogether.

> Just one small thing: should the small horizontal "jump" of menuitems with the check icon be fixed in this PR?

This bug exists in `master`, so I'll create a separate PR to fix it. I'm keen to get this PR merged as it is blocking some other work.
